File: BojS2_1012.py
# 유기농 배추
import sys
sys.setrecursionlimit(10 ** 4)  # O(n*m) = 2500


def solution():
    for _ in range(k):
        y, x = map(int, input().split())
        graph[x][y] = 1

    result = 0
    for i in range(n):
        for j in range(m):
            if dfs(i, j):
                result += 1
    print(result)


def dfs(x, y):
    if x <= -1 or x >= n or y <= -1 or y >= m:
        return False

    if graph[x][y] == 1:
        graph[x][y] = 0
        dfs(x - 1, y)
        dfs(x, y - 1)
        dfs(x + 1, y)
        dfs(x, y + 1)
        return True
    return False


if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        m, n, k = map(int, input().split())
        graph = [[0] * m for _ in range(n)]
        solution()

File: BojS2_1012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BojS2_1012 {   // 유기농 배추

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;
    int m;
    int n;
    int k;
    int[][] graph;

    public BojS2_1012() throws IOException {
        st = new StringTokenizer(br.readLine());
        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
    }

    private void solution() throws IOException {
        graph = new int[n][m];
        createGraph(k);
        int result = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (dfs(i, j)) {
                    result++;
                }
            }
        }
        System.out.println(result);
    }

    private void createGraph(int k) throws IOException {
        for (int i = 0; i < k; i++) {
            st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());
            graph[y][x] = 1;
        }
    }

    private boolean dfs(int x, int y) {
        if (x >= 0 && x < n && y >= 0 && y < m && graph[x][y] == 1) {
            graph[x][y] = 0;
            dfs(x, y - 1);
            dfs(x - 1, y);
            dfs(x, y + 1);
            dfs(x + 1, y);
            return true;
        }
        return false;
    }

    public static void main(String[] args) throws IOException {
        int t = Integer.parseInt(br.readLine());
        for (int i = 0; i < t; i++) {
            BojS2_1012 problem = new BojS2_1012();
            problem.solution();
        }
    }

}
